src/models/rule.rs

```diff
@@ -198,11 +198,15 @@ pub(crate) struct InstantiatedRule {
 
 impl InstantiatedRule {
   pub(crate) fn new(rule: &Rule, substitutions: &HashMap<String, String>) -> Self {
-    let substitutions_for_holes = rule
+    let substitutions_for_holes: HashMap<String, String> = rule
       .holes()
       .iter()
       .filter_map(|h| substitutions.get(h).map(|s| (h.to_string(), s.to_string())))
       .collect();
+    if substitutions_for_holes.len() != rule.holes().len() {
```

Comment:
Again, question, what happens if the lengths match, but not the keys? Wouldn't you still end with a partially instantiated rule? Do we want to prevent this/check for it?

Reply:
As `substitutions`  we pass the entire substitution map we have in the "state". 
Then in this statement: 
```
let substitutions_for_holes: HashMap<String, String> = rule
      .holes()
      .iter()
      .filter_map(|h| substitutions.get(h).map(|s| (h.to_string(), s.to_string())))
      .collect();
```
I iterate over each hole and get its value from `substitutions` parameter map if possible, and collect. 
So when lengths don't match, it implies that we did not find all substitutions. 
If lengths match, implies We found a substitution for each hole in the rule (note we iterate over `rule.holes())
So to answer your question it will never happen that rule is partially instantiated though the lengths match

Reply:
Ok, I get it now. Fair enough!

I do feel this logic is complex enough that it merits a comment in the code. Probably not the full long explanation you just gave, but a sentence or two. Maybe:

```
// filter_map discards any element of `rules.holes()` for which there isn't a valid substitution,
// so checking that the lengths match is enough to verify all holes have a matching substitution.
```

As a side note, what goes Rust do on `m.get(k).map(...)` where `k` is missing from `m`? My intuition from Java/Go is that this would be an NPE/nil panic because `m.get(k)` returns `null`/`nil`. But presumably that isn't the case in Rust?

Do we have a unit test that expects this panic? I don't immediately see a test case for invalid instantiation in this PR, but not sure if we already had it before.